docs/design/per-op-config.md

```diff
@@ -0,0 +1,179 @@
+    }
+    
+    launch {
+        doBackgroundRoutine(client.withConfig {
+            logMode = SdkLogMode.LogResponse
+        })
+        
+        doBackgroundRoutine2(client.withConfig {
+            httpEngine = YourHttpClientEngineImpl()
+        })
+    }
+    
+    // application continues...
+}
+```
+
+Some observations about the above sample:
+* Since no explicit HTTP engine is passed to construction of the root client, the SDK creates its own "managed" default.
+* The first use of `withConfig` and `use` demonstrates how to make basic one-off calls. The defaulted HTTP engine is
+  inherited by the child. When the service client is `closed()` as part of the `use { }` call, the runtime recognizes
+  that others are still holding onto its HTTP engine and so it isn't fully closed.
+* The client passed to `doBackgroundRoutine` inherits the default HTTP engine as well. Both the parent and subclient can
+  be closed at any time in either order - the managed HTTP engine will only be reaped when both clients using it are
+  closed.
+* The user passes their own HTTP engine implementation in the `doBackgroundRoutine2` call, and thus is responsible for
+  closing it separately from the client.
+
+# rejected alternatives
+
+## alt1: request object builder
+Add an sdk-specific field on request builders (proposed name `withConfig` used in this sample). Client implementations
+retrieve that builder and apply it to the config on specific requests.
+
+```kotlin
+// showing definition of sample client operationA input
+class OperationARequest private constructor(builder: Builder) {
+    val requestValue: Int? = builder.requestValue
+    // ...
+    class Builder {
+        var requestValue: Int? = null
+        // ...
+        fun withConfig(configBuilder: ServiceClient.Config.Builder.() -> Unit) { /* ... */ }
+    }
+}
+```
+
+Sample usage:
+```kotlin
+suspend fun main() {
+    ServiceClient().use {
+        it.operationA {
+            requestValue = 1
+        }
+        it.operationA {
+            requestValue = 2
+            withConfig {
+                logMode = SdkLogMode.LogRequestWithBody
+            }
+        }
+    }
+}
+```
+
+Rejected because:
+* Adding our own named field `withConfig` to a structure generated based on a model we don't own introduces a collision
+  risk. There's nothing preventing a model author from adding their own "withConfig" field to a shape.
+* Affords less flexibility. A caller may with to structure a series of operations as a function which accepts an
+  arbitrarily configured client - this route removes the possibility for that separation, forcing the caller to hardcode
+  config overrides in specific method calls.
+* Prevents us from supporting "shape-only" (generate only client interface and input/output shapes, no implementation)
+  codegen in the future, since the `withConfig` property requires additional mechanisms to be generated into the structs
+  in order to be retrieved by the implemented operation.
```

Comment:
**Nit**: "Prevents us from" → "Complicates". There are paths forward here that still allow shape-only codegen but they require complicating our generators.

Reply:
I agree with your statement but I think it's even more than that though.

Generating request or response structures with additional client level configuration makes it almost impossible to re-use "shape only" codegen.

e.g. Let's say we have wanted to support shape only codegen and generate just the request/response structures (no serde, no client interface). Theoretically it should be possible then to depend on that package and use those types as the inputs to a service client operation. Tying them together this way would make them incompatible though.